src/activity_tracker.cpp

```diff
@@ -0,0 +1,136 @@
+{
+    if( intake > tracker ) {
+        return tracker * 0.5;
+    }
+    return tracker - intake * 0.5;
+}
+
+// Called every 5 minutes, when activity level is logged
+void activity_tracker::try_reduce_weariness( int bmr, bool sleeping )
+{
+    tick_counter++;
+    if( average_activity() - NO_EXERCISE <= std::numeric_limits<float>::epsilon() ) {
+        low_activity_ticks++;
+        // Recover twice as fast at rest
+        if( sleeping ) {
+            low_activity_ticks++;
+        }
+    }
+
+    const float recovery_mult = get_option<float>( "WEARY_RECOVERY_MULT" );
+
+    if( low_activity_ticks >= 6 ) {
+        int reduction = tracker;
+        // 1/20 of whichever's bigger
+        if( bmr > reduction ) {
+            reduction = bmr * recovery_mult;
+        } else {
+            reduction *= recovery_mult;
+        }
+        low_activity_ticks -= 6;
+
+        tracker -= reduction;
+    }
+
+    if( tick_counter >= 12 ) {
+        intake *= 1 - recovery_mult;
+        tick_counter -= 12;
+    }
+
+    // Normalize values, make sure we stay above 0
+    intake = std::max( intake, 0 );
+    tracker = std::max( tracker, 0 );
+    tick_counter = std::max( tick_counter, 0 );
+    low_activity_ticks = std::max( low_activity_ticks, 0 );
+}
+
+void activity_tracker::weary_clear()
+{
+    tracker = 0;
+    intake = 0;
+    low_activity_ticks = 0;
+    tick_counter = 0;
+}
+
+std::string activity_tracker::debug_weary_info() const
+{
+    return string_format( "Intake: %d Tracker: %d", intake, tracker );
+}
+
+void activity_tracker::calorie_adjust( int nkcal )
+{
+    if( nkcal > 0 ) {
+        intake += nkcal;
+    } else {
+        // nkcal is negative, we need positive
+        tracker -= nkcal;
+    }
+}
+
+float activity_tracker::activity() const
+{
+    if( current_turn == calendar::turn ) {
+        return current_activity;
+    }
+    return 1.0f;
+}
+
+float activity_tracker::average_activity() const
+{
+    if( activity_reset && current_turn != calendar::turn ) {
+        return previous_activity / num_events;
+    }
+    return ( accumulated_activity + current_activity ) / num_events;
+}
+
+// The idea here is the character is going about their business logging activities,
+// and log_activity() handles sorting them out, it records the largest magnitude for a given turn,
+// and then rolls the previous turn's value into the accumulator once a new activity is logged.
+// After a reset, we have to pretend the previous values weren't logged.
+void activity_tracker::log_activity( float new_level )
+{
+    if( current_turn == calendar::turn ) {
+        current_activity = std::max( current_activity, new_level );
+    } else if( activity_reset ) {
+        activity_reset = false;
+        current_activity = new_level;
+        accumulated_activity = 0.0f;
+        num_events = 1;
+    } else {
+        accumulated_activity += current_activity;
+        // skipped turns?
+        current_activity = new_level;
+        num_events++;
```

Comment:
I think this should look like this to account for the turns thing
```suggestion
        int turns_moved = calendar::turn - current_turn;
        accumulated_activity += current_activity * turns_moved;
        // skipped turns?
        current_activity = new_level;
        num_events += turns_moved;
```

Reply:
Sort of, we can do number of turns, but the activity level if anything should be NO_EXERCISE, representative of the fact that no activity was logged in the interim.

Reply:
I'd have to look through to see when skipped turns trigger, but this is the correct behavior for what I was considering, when an activity takes more than a characters moves to complete. E.g. if I drag furniture and it takes me 6 turns (but it is not an activity), I should be counted as whatever activity level for 6 turns, instead of just 1, and nothing for 5 turns.

Reply:
The activity should be logging its exertion level every turn.

Reply:
I am talking about non-activity actions like that. Dragging is now an activity, but if it were not this would be the correct behavior.
I am not sure how many of these we have, however. Melee combat comes to mind as somewhere this could pop up.

Reply:
Still, anything the player does should log an activity, and anything that takes more than one turn should log each turn.
If that isn't happening, we can fix it.